### PR TITLE
CYS - reduce editor instance re-render

### LINKED ISSUE #45420: CYS: Header/Footer pattern aren't rendered on the live preview
In some cases, the selected header or footer pattern isn't rendered on the live preview. I'm able to replicate this issue with the cache disabled.

1. Ensure that cache is disabled.
2. Switch the header/footer patterns until the select header or footer isn't visible anymore (check the video).

https://github.com/woocommerce/woocommerce/assets/4463174/44acc81a-1b49-4483-a1a5-3878225e54cf

### LINKED ISSUE #45448: [CYS on Core] After first completing the assembler hub, subsequent uses can lead to a blank transitional page at the end.
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [ ] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

If a user accesses the assembler hub, clicks Done and lands on the final page, subsequent usage of the hub will lead to this URL when finishing: `/wp-admin/admin.php?page=wc-admin&customizing=true&path=%2Fcustomize-store%2Ftransitional`

Upon reloading the page or accessing that URL again, it's blank.

### Expected behavior

Unsure of the exact expected behaviour, but not a blank page. I believe the final URL should land on `/wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store%2Fassembler-hub`.

### Actual behavior

Blank page and JS errors:

```
TypeError: can't access property "id", o is null
    jo customize-store.js:1
    St react-dom.min.js:1
    $s react-dom.min.js:1
    Sl react-dom.min.js:1
    kl react-dom.min.js:1
    bl react-dom.min.js:1
    dl react-dom.min.js:1
    Nn react-dom.min.js:1
    ul react-dom.min.js:1
    Dt react-dom.min.js:1
    Ft react-dom.min.js:1
    s index.js:89
    c index.js:96
    r registry.js:247
    Redux 4
    F promise-middleware.js:20
    U resolvers-cache-middleware.js:45
    s Redux
```

### Steps to reproduce

1. Install the prerequisites to test CYS.
2. Access CYS and click Start Designing
3. Make a change to any section and save it. (Optional, but useful for remembering if you've accessed it)
4. Click Done. You'll land on this URL: `/wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store%2Fassembler-hub`
5. Access it again and click Done.
6. You'll land on this URL `/wp-admin/admin.php?page=wc-admin&customizing=true&path=%2Fcustomize-store%2Ftransitional`
7. Reload it to see a blank page and JS errors.

### WordPress Environment

```
### WordPress Environment ###

WordPress address (URL): https://dangerously-nutty-kid.jurassic.ninja
Site address (URL): https://dangerously-nutty-kid.jurassic.ninja
WC Version: 8.8.0
REST API Version: ✔ 8.8.0
Action Scheduler Version: ✔ 3.7.2
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: ✔

### Server Environment ###

Server Info: nginx
PHP Version: 8.1.27
PHP Post Max Size: 2 GB
PHP Time Limit: 1200
PHP Max Input Vars: 6144
cURL Version: 8.4.0
OpenSSL/1.1.1w

SUHOSIN Installed: –
MySQL Version: 10.6.15-MariaDB-log
Max Upload Size: 2 GB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.8.0
WC Database Prefix: wp_
Total Database Size: 3.04MB
Database Data Size: 1.37MB
Database Index Size: 1.67MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 0.09MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.22MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 8
page: 7
post: 2
product: 6
revision: 11
wp_global_styles: 1
wp_navigation: 1
wp_template: 1
wp_template_part: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (5) ###

Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.1
Companion Plugin: by Osk – 1.30
Gutenberg: by Gutenberg Team – 17.8.2
Jetpack: by Automattic – 13.2-beta
WooCommerce: by Automattic – 8.8.0-dev

### Inactive Plugins (0) ###


### Dropin Plugins () ###

advanced-cache.php: advanced-cache.php
object-cache.php: Memcached

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 231 B

### WC Pages ###

Shop base: #8 - /shop/
Cart: #9 - /cart/
Checkout: #10 - /checkout/
My account: #11 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Four
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ❌ Not declared

### Templates ###

Overrides: /srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/error.php
/srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/notice.php
/srv/htdocs/wp-content/plugins/woocommerce/templates/block-notices/success.php


### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field
launch-your-store

Daily Cron: ✔ Next scheduled: 2024-03-09 16:46:44 +00:00
Options: ✔
Notes: 66
Onboarding: skipped

### Action Scheduler ###

Complete: 8
Oldest: 2024-03-08 16:48:09 +0000
Newest: 2024-03-08 16:48:09 +0000

Pending: 1
Oldest: 2024-03-09 16:48:09 +0000
Newest: 2024-03-09 16:48:09 +0000


### Status report information ###

Generated at: 2024-03-08 17:10:29 +00:00
```

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by gigitux**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #45420.
Partially fixes #45448.

### How to test the changes in this Pull Request:

This PR request aims to optimize the block editor by minimizing unnecessary re-renders. The current issue involves suboptimal performance and the occurrence of unexpected bugs, mainly due to the block repeatedly re-instantiating itself. In response, I've refactored the code by centralizing the logic within a parent component, effectively streamlining the block editor component and enhancing overall efficiency. This change is crucial for maintaining a smoother editing experience and addressing potential issues stemming from excessive re-renders.

I think that there is space for more optimization, but at least for now, I don't think that it is worth spending time on it!

| Before | After |
|--------|--------|
|<video src=https://github.com/woocommerce/woocommerce/assets/4463174/d6ae11d6-910a-42d2-a643-4bc39c726f89 /> | <video src=https://github.com/woocommerce/woocommerce/assets/4463174/9f01de92-ce43-4344-b62e-57a7fbe2161f /> | 

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new WooCommerce installation with this version.
2. Ensure the WooCommerce Beta Tester plugin is installed and activated (available on this monorepo).
3. Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag.
4. Install the last version of Gutenberg.
5. Visit the `wp-admin/admin.php?page=wc-admin&path=/customize-store`.
10. Open the network tools.
11. Disable the cache.
12. Follow the process.
13. Switch the header/footer patterns fastly (check #45420 to understand the issue)
14. Ensure that the pattern is still visible.
15. Click Done. You'll land on this URL: `/wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store%2Fassembler-hub`
16. Access it again and click Done.
17. You'll land on this URL `/wp-admin/admin.php?page=wc-admin&customizing=true&path=%2Fcustomize-store%2Ftransitional`.
18. Ensure that the the transitional page is visible. 

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

CYS - reduce editor instance re-render.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-admin/client/customize-store/assembler-hub/block-editor.tsx
Comment (gigitux) [inline plugins/woocommerce-admin/client/customize-store/assembler-hub/block-editor.tsx:1]: I just moved the delete lines to `block-editor-container.tsx` file.

**Comment by nefeline** [comment]
Thanks for working on this @gigitux !

> This PR request aims to optimize the block editor by minimizing unnecessary re-renders. The current issue involves suboptimal performance and the occurrence of unexpected bugs, mainly due to the block repeatedly re-instantiating itself. In response, I've refactored the code by centralizing the logic within a parent component, effectively streamlining the block editor component and enhancing overall efficiency. This change is crucial for maintaining a smoother editing experience and addressing potential issues stemming from excessive re-renders.

✅ I can confirm the Pattern Assembler works as expected with this change in place, although I still do see JS errors whenever attempting to switch in-between the header options:

Screenshot 2024-03-11 at 16 47 31

I can reproduce this error both on production and on this PR, so it's unrelated, but we should address it eventually. Created https://github.com/woocommerce/woocommerce/issues/45486 so we can work on it separately.

> I think that there is space for more optimization, but at least for now, I don't think that it is worth spending time on it!

Can you share more details about the other optimizations you believe would be suitable for implementing here? Do we have any issues created for this purpose already?

✅ I can also confirm that the problem with the transitional page, as reported on #45448, has been partially fixed: the only piece still pending is to ensure that whenever the user clicks on Done when going through the CYS flow for the very first time, they are led to `wp-admin/admin.php?page=wc-admin&customizing=true&path=%2Fcustomize-store%2Ftransitional` instead of keeping the URL unchanged `wp-admin/admin.php?page=wc-admin&path=%2Fcustomize-store%2Fassembler-hub`. As previously discussed, we can postpone working on this considering we do have other high priority issues to tackle first.

**Comment by gigitux** [comment]
> Can you share more details about the other optimizations you believe would be suitable for implementing here? Do we have any issues created for this purpose already?

Unfortunally I didn't have time to check and go in-depth, so I don't have much to say. Mostly, I noticed that there are some hooks that we import from Gutenberg that trigger the re-render of the component multiple times. "re-renders" doesn't mean bad performance, but still, I'm curious to understand why they are necessary. It could make sense to open an issue as "explorative work," but I would not focus on this in the short/mid-term. IMHO, it makes sense to investigate performance when:

- the app is slow
- weird workaround are necessary (like this: https://github.com/woocommerce/woocommerce/pull/45399)


If you agree with the changes, could you approve the PR so I can merge it? Thanks!

**Review by nefeline — APPROVED** [review]
> Unfortunally I didn't have time to check and go in-depth, so I don't have much to say. Mostly, I noticed that there are some hooks that we import from Gutenberg that trigger the re-render of the component multiple times. "re-renders" doesn't mean bad performance, but still, I'm curious to understand why they are necessary. It could make sense to open an issue as "explorative work," but I would not focus on this in the short/mid-term.

Gotcha! Thanks for clarifying: I noticed no issue was created, but I still strongly recommend doing so, as findings such as the one you just shared here are invaluable information that can help us improve the health and/or performance of our codebase and application in the long term.

> If you agree with the changes, could you approve the PR so I can merge it? Thanks!

Absolutely, thanks for your work here!

**Comment by gigitux** [comment]
Thanks for the feedback and for the review! Issue created: https://github.com/woocommerce/woocommerce/issues/45517

**Pull request merged**